src/TemplateScope.ts

```diff
@@ -3,50 +3,96 @@
 // ----------------------------------------------------------------------------
 
 import * as assert from 'assert';
-import { Utilities } from '../extension.bundle';
+import { Json, Utilities } from '../extension.bundle';
+import { CachedValue } from './CachedValue';
 import { templateKeys } from './constants';
 import { IParameterDefinition } from "./IParameterDefinition";
+import { IResource } from './IResource';
 import * as TLE from "./TLE";
 import { UserFunctionDefinition } from './UserFunctionDefinition';
 import { UserFunctionNamespaceDefinition } from "./UserFunctionNamespaceDefinition";
 import { IVariableDefinition } from './VariableDefinition';
 
-export enum ScopeContext {
+export enum TemplateScopeKind {
     TopLevel = "TopLevel",
     ParameterDefaultValue = "ParameterDefaultValue",
-    UserFunction = "UserFunction"
+    UserFunction = "UserFunction",
+    NestedDeploymentWithInnerScope = "NestedDeploymentWithInnerScope",
+    NestedDeploymentWithOuterScope = "NestedDeploymentWithOuterScope",
 }
 
 /**
  * Represents the scoped access of parameters/variables/functions at a particular point in the template tree.
  */
-export class TemplateScope {
-    /**
-     * Constructor
-     */
+export abstract class TemplateScope {
+    private _parameterDefinitions: CachedValue<IParameterDefinition[] | undefined> = new CachedValue<IParameterDefinition[] | undefined>();
+    private _variableDefinitions: CachedValue<IVariableDefinition[] | undefined> = new CachedValue<IVariableDefinition[] | undefined>();
+    private _functionDefinitions: CachedValue<UserFunctionNamespaceDefinition[] | undefined> = new CachedValue<UserFunctionNamespaceDefinition[] | undefined>();
+    private _resources: CachedValue<IResource[] | undefined> = new CachedValue<IResource[] | undefined>();
+
+    public readonly abstract scopeKind: TemplateScopeKind;
```

Comment:
>TemplateScopeKind [](start = 40, length = 17)

Do all fields (all definition kinds) make sense for all TemplateScopeKind values?

Reply:
Need to be available to be queried, yes (some may be unsupported)